server/services/visitNotificationsService.ts

```diff
@@ -16,32 +19,123 @@ export default class VisitNotificationsService {
+
+    const type = new Map<string, string>()
+    const typeFilterItems: FilterField['items'] = []
+
+    notificationGroups.forEach(notificationGroup =>
+      type.set(notificationGroup.type, notificationTypes[notificationGroup.type]),
+    )
+    type.forEach((label, value) => typeFilterItems.push({ label, value, checked: appliedFilters.type.includes(value) }))
+    typeFilterItems.sort((a, b) => a.label.localeCompare(b.label))
+
+    return [
+      {
+        id: 'bookedBy',
+        label: 'Booked by',
+        items: bookedByFilterItems,
+      },
+      {
+        id: 'type',
+        label: 'Reason',
+        items: typeFilterItems,
+      },
+    ]
+  }
+
+  private cleanAppliedFilters(appliedFilters: AppliedFilters, filters: FilterField[]): AppliedFilters {
+    const validUsernames = filters.find(filter => filter.id === 'bookedBy')?.items.map(item => item.value) || []
+    const validTypes = filters.find(filter => filter.id === 'type')?.items.map(item => item.value) || []
+
+    const appliedFiltersCleaned = {
+      bookedBy: appliedFilters.bookedBy.filter(user => validUsernames.includes(user)),
+      type: appliedFilters.type.filter(type => validTypes.includes(type)),
+    }
+
+    return appliedFiltersCleaned
+  }
+
+  private buildVisitsReviewListItem(notificationGroup: NotificationGroup): VisitsReviewListItem {
+    const bookedByNames = notificationGroup.affectedVisits.map(visit => visit.bookedByName)
+
+    const prisonerNumbers =
+      notificationGroup.type === 'NON_ASSOCIATION_EVENT'
+        ? notificationGroup.affectedVisits.map(visit => visit.prisonerNumber)
+        : [notificationGroup.affectedVisits[0].prisonerNumber]
+
+    const visitDates =
+      notificationGroup.type === 'NON_ASSOCIATION_EVENT'
+        ? [prisonerDateTimePretty(notificationGroup.affectedVisits[0].visitDate)]
+        : notificationGroup.affectedVisits.map(visit => prisonerDateTimePretty(visit.visitDate))
+
+    return {
+      bookedByNames,
+      prisonerNumbers,
+      reference: notificationGroup.reference,
+      type: notificationGroup.type,
+      visitDates,
+    }
+  }
+
+  private filterNotificationGroups(
+    appliedFilters: AppliedFilters,
+    notificationGroups: NotificationGroup[],
+  ): NotificationGroup[] {
+    if (!Object.values(appliedFilters).flat().length) return notificationGroups
+
+    return notificationGroups.filter(notificationGroup => {
+      const notificationGroupUsers = notificationGroup.affectedVisits.flatMap(visit => visit.bookedByUserName)
+      const isUsernameMatch = appliedFilters.bookedBy.length
+        ? appliedFilters.bookedBy.some(user => notificationGroupUsers.includes(user))
+        : true
+
+      const isTypeMatch = appliedFilters.type.length
+        ? appliedFilters.type.some(type => type === notificationGroup.type)
+        : true
+
+      return isUsernameMatch && isTypeMatch
+    })
+  }
+
+  async getVisitsReviewList(
```

Comment:
A fair bit going on here; now split out to different private helper methods. API just gives all data so identifying potential filter values and actually filtering all done here. 

If in practice there ends up being a lot of data we can revisit and do more work (inc pagination) in the API - but this was discounted for initial release.